Plugin.php

```diff
@@ -8,28 +8,192 @@
 namespace yii\composer;
 
 use Composer\Composer;
+use Composer\DependencyResolver\Operation\UpdateOperation;
+use Composer\EventDispatcher\EventSubscriberInterface;
+use Composer\Installer\PackageEvent;
+use Composer\Installer\PackageEvents;
 use Composer\IO\IOInterface;
 use Composer\Plugin\PluginInterface;
+use Composer\Script;
+use Composer\Script\ScriptEvents;
 
 /**
  * Plugin is the composer plugin that registers the Yii composer installer.
  *
  * @author Qiang Xue <qiang.xue@gmail.com>
  * @since 2.0
  */
-class Plugin implements PluginInterface
+class Plugin implements PluginInterface, EventSubscriberInterface
 {
     /**
+     * @var array noted package updates.
+     */
+    private $_packageUpdates = [];
```

Comment:
we **do** name all private properties with `_` prefix to have a simple rule: "Private variables should be named like $_varName." instead of introducing a complex rule that lists the cases where it is allowed and where not. I do not care if our code style is 100% compatible to PSR. Lets make it practical instead of introducing complex rules that none can remember.